encoding/ssh/filexfer/buffer.go

```diff
@@ -0,0 +1,268 @@
+func (b *Buffer) ConsumeUint32() (uint32, error) {
+	if b.Len() < 4 {
+		return 0, ErrShortPacket
+	}
+
+	v := binary.BigEndian.Uint32(b.b[b.off:])
+	b.off += 4
+	return v, nil
+}
+
+// AppendUint32 appends a single uint32 into the Buffer, in network byte order (big-endian).
+func (b *Buffer) AppendUint32(v uint32) {
+	b.b = append(b.b,
+		byte(v>>24),
+		byte(v>>16),
+		byte(v>>8),
+		byte(v>>0),
+	)
+}
+
+// ConsumeUint64 consumes a single uint64 from the Buffer, in network byte order (big-endian).
+// If Buffer does not have enough data, it will return ErrShortPacket.
+func (b *Buffer) ConsumeUint64() (uint64, error) {
+	if b.Len() < 8 {
+		return 0, ErrShortPacket
+	}
+
+	v := binary.BigEndian.Uint64(b.b[b.off:])
+	b.off += 8
+	return v, nil
+}
+
+// AppendUint64 appends a single uint64 into the Buffer, in network byte order (big-endian).
+func (b *Buffer) AppendUint64(v uint64) {
+	b.b = append(b.b,
+		byte(v>>56),
+		byte(v>>48),
+		byte(v>>40),
+		byte(v>>32),
+		byte(v>>24),
+		byte(v>>16),
+		byte(v>>8),
+		byte(v>>0),
+	)
+}
+
+// ConsumeInt64 consumes a single int64 from the Buffer, in network byte order (big-endian) with two’s complement.
+// If Buffer does not have enough data, it will return ErrShortPacket.
+func (b *Buffer) ConsumeInt64() (int64, error) {
+	u, err := b.ConsumeUint64()
+	if err != nil {
+		return 0, err
+	}
+
+	return int64(u), err
+}
+
+// AppendInt64 appends a single uint64 into the Buffer, in network byte order (big-endian) with two’s complement.
+func (b *Buffer) AppendInt64(v uint64) {
+	b.AppendUint64(uint64(v))
+}
+
+// ConsumeByteSlice consumes a single string of raw binary data from the Buffer.
+// A string is a uint32 length, followed by that number of raw bytes.
+// If Buffer does not have enough data, or defines a length larger than available, it will return ErrShortPacket.
+func (b *Buffer) ConsumeByteSlice() ([]byte, error) {
+	length, err := b.ConsumeUint32()
+	if err != nil {
+		return nil, err
+	}
+
+	if b.Len() < int(length) {
+		return nil, ErrShortPacket
+	}
+
+	v := b.b[b.off:]
+	if len(v) > int(length) {
+		v = v[:length:length]
+	}
+	b.off += int(length)
+	return v, nil
+}
+
+// AppendByteSlice appends a single string of raw binary data into the Buffer.
+// A string is a uint32 length, followed by that number of raw bytes.
+func (b *Buffer) AppendByteSlice(v []byte) {
+	b.AppendUint32(uint32(len(v)))
+	b.b = append(b.b, v...)
+}
+
+// ConsumeString consumes a single string of UTF-8 encoded text from the Buffer.
+// A string is a uint32 length, followed by that number of raw bytes.
+// If Buffer does not have enough data, or defines a length larger than available, it will return ErrShortPacket.
+func (b *Buffer) ConsumeString() (string, error) {
+	v, err := b.ConsumeByteSlice()
+	if err != nil {
+		return "", err
+	}
+
+	return string(v), nil
+}
+
+// AppendString appends a single string of UTF-8 encoded text into the Buffer.
```

Comment:
Nothing in the code says UTF-8, nor should it. Go doesn't care what we put in our strings and the filexfer draft only specifies the encoding of error messages.

Reply:
The Go standard actually does say that the string type is “implicitly UTF-8 encoded”. Specifically, it is defined behavior that a `range` over a string will interpret that string as UTF-8 runes, and those runes will be iterated through, not the actual bytes. Therefore, I took the choice that anything intended to fully arbitrary data should be handled (as per Go Best Practice) a byte slice, while respecting that strings should be treated as very-likely-to-be-turned-into-UTF-8.

While the filexfer-02 standard does not explicitly say that most strings are UTF-8 encoded, starting from filexfer-03 many of the string fields _are_ defined to be UTF-8 only (unless otherwise negotiated), the only ones that stay as arbitrary byte encodings are on the FXP_DATA, and FXP_WRITE packets, as well as all of the handles.

I think I will however update the documentation by saying that due to how Go works, strings are necessarily always implicitly UTF-8, and “as per usual” care must be taken when using them for arbitrary data or other encodings.